v-next/hardhat-viem-matchers/src/internal/matchers/revert/utils.ts

```diff
@@ -0,0 +1,39 @@
+import type { GenericFunction } from "../../../types.js";
+
+import assert from "node:assert/strict";
+
+import { assertHardhatInvariant } from "@nomicfoundation/hardhat-errors";
+import { ensureError } from "@nomicfoundation/hardhat-utils/error";
+
+export async function checkRevert(fn: GenericFunction): Promise<string> {
+  let hasReverted = false;
+  let reason = "";
+
+  try {
+    await fn();
+  } catch (error) {
+    ensureError(error);
+
+    if ("details" in error) {
+      hasReverted = true;
+
+      assertHardhatInvariant(
+        typeof error.details === "string",
+        "The error details should be a string",
+      );
+
+      reason = error.details
+        .split("reverted with reason string ")[1]
+        .replaceAll("'", "");
```

Comment:
I don't think this is very future-proof. Ideally we want an API-stable way of getting the return data, and then we decode it ourselves. I think this will be tricky, for fun JSON-RPC reasons, but it's important.